File: src/main/java/Day9.java
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Day9 {

    public static void main(String[] args) throws Exception {
        solve();
    }

    private static void solve() throws Exception {
        try (Stream<String> stream = Files.lines(Paths.get(Day9.class.getResource("day9.txt").toURI()))) {
            List<String> lines = stream.toList();

            int resultPart1 = 0;
            int resultPart2 = 0;
            for (String line : lines) {
                LinkedList<Integer> base = Arrays.stream(line.split(" "))
                        .map(Integer::parseInt)
                        .collect(Collectors.toCollection(LinkedList::new));

                List<LinkedList<Integer>> rows = buildRowsForOneLine(base);
                fillInValues(rows);
                LinkedList<Integer> topRow = rows.get(rows.size() - 1);
                resultPart1 += topRow.getLast();
                resultPart2 += topRow.getFirst();
            }
            System.out.println(resultPart1);
            System.out.println(resultPart2);
        }
    }

    private static void fillInValues(List<LinkedList<Integer>> rows) {
        Collections.reverse(rows);
        for (int i = 0; i < rows.size(); i++) {
            fillInNextValue(rows, i);
            fillInFirstValue(rows, i);
        }
    }

    private static void fillInNextValue(List<LinkedList<Integer>> rows, int i) {
        int lastPrev = i != 0 ? rows.get(i - 1).getLast() : 0;
        int lastCurr = rows.get(i).getLast();
        int newLastCurr = lastCurr + lastPrev;
        rows.get(i).addLast(newLastCurr);
    }

    private static void fillInFirstValue(List<LinkedList<Integer>> rows, int i) {
        int firstPrev = i != 0 ? rows.get(i - 1).get(0) : 0;
        int firstCurr = rows.get(i).getFirst();
        int newLastCurr = firstCurr - firstPrev;
        rows.get(i).addFirst(newLastCurr);
    }

    private static List<LinkedList<Integer>> buildRowsForOneLine(LinkedList<Integer> base) {
        List<LinkedList<Integer>> rows = new ArrayList<>();
        rows.add(base);
        int rowIndex = 0;

        while (true) {
            LinkedList<Integer> newRow = new LinkedList<>();
            LinkedList<Integer> prevRow = rows.get(rowIndex);
            for (int i = 1; i < prevRow.size(); i++) {
                int prev = prevRow.get(i - 1);
                int curr = prevRow.get(i);
                newRow.add(curr - prev);
            }
            rows.add(newRow);
            if (newRow.stream().allMatch(nbr -> nbr == 0)) {
                break;
            }
            rowIndex++;
        }
        return rows;
    }

}

File: python/day9.py
file = open('./../src/main/resources/day9.txt', 'r')
lines = file.readlines()


def build_rows(nbr_line):
    result = [nbr_line]
    row_index = 0
    while True:
        prev_row = result[row_index]
        new_row = []
        for i in range(1, len(prev_row)):
            num1 = prev_row[i - 1]
            num2 = prev_row[i]
            new_row.append(num2 - num1)
        result.append(new_row)
        if all(nbr == 0 for nbr in new_row):
            break
        row_index += 1
    return result


def part1():
    result = 0
    for line in lines:
        rows = build_rows([int(num) for num in line.split()])
        rows.reverse()
        for index, row in enumerate(rows):
            last_elem_prev_row = rows[index - 1][-1] if index != 0 else 0
            last_elem_curr_row = row[-1]
            row.append(last_elem_curr_row + last_elem_prev_row)
        top_row = rows[-1]
        result += top_row[-1]
    return result


print(part1())
